internal/impl/sftp/input.go

```diff
@@ -173,184 +173,190 @@ func newSFTPReaderFromParsed(conf *service.ParsedConfig, mgr *service.Resources)
 		if ctx.Err() != nil {
 			return nil, nil, ctx.Err()
 		}
-		_ = scanner.Close(ctx)
-		s.scannerMut.Lock()
-		if s.currentPath == currentPath {
-			s.scanner = nil
-			s.currentPath = ""
-		}
-		s.scannerMut.Unlock()
+		_ = s.scanner.Close(ctx)
+		s.scanner = nil
+		s.currentPath = ""
 		if errors.Is(err, io.EOF) {
 			err = service.ErrNotConnected
 		}
 		return nil, nil, err
 	}
 
 	for _, part := range parts {
-		part.MetaSetMut("sftp_path", currentPath)
+		part.MetaSetMut("sftp_path", s.currentPath)
 	}
 
-	return parts, func(ctx context.Context, res error) error {
-		return codecAckFn(ctx, res)
-	}, nil
+	return parts, codecAckFn, nil
 }
 
-func (s *sftpReader) Close(ctx context.Context) error {
-	s.scannerMut.Lock()
+func (s *sftpReader) initScanner(ctx context.Context) error {
+	s.stateLock.Lock()
 	scanner := s.scanner
-	s.scanner = nil
-	client := s.client
-	s.client = nil
-	s.paths = nil
-	s.scannerMut.Unlock()
-
+	s.stateLock.Unlock()
 	if scanner != nil {
-		if err := scanner.Close(ctx); err != nil {
-			s.log.With("error", err).Warn("Failed to close consumed file")
+		return nil
+	}
+
+	var file *sftp.File
+	var path string
+	for {
+		var ok bool
+		var err error
+		path, ok, err = s.pathProvider.Next(ctx)
+		if err != nil {
+			return fmt.Errorf("finding next file path: %w", err)
+		}
+		if !ok {
+			return service.ErrEndOfInput
+		}
+
+		file, err = s.client.Open(path)
+		if err != nil {
+			s.log.With("path", path, "err", err.Error()).Warn("Unable to open previously identified file")
+			if os.IsNotExist(err) {
+				// If we failed to open the file because it no longer exists then we
+				// can "ack" the path as we're done with it. Otherwise we "nack" it
+				// with the error as we'll want to reprocess it again later.
+				err = nil
+			}
+			if ackErr := s.pathProvider.Ack(ctx, path, err); ackErr != nil {
+				s.log.With("error", ackErr).Warnf("Failed to acknowledge path: %s", path)
+			}
+			continue
+		}
+
+		details := service.NewScannerSourceDetails()
+		details.SetName(path)
+		scanner, err := s.scannerCtor.Create(file, s.newCodecAckFn(path), details)
+		if err != nil {
+			_ = file.Close()
+			return fmt.Errorf("creating scanner: %w", err)
 		}
+
+		s.stateLock.Lock()
+		s.scanner = scanner
+		s.currentPath = path
+		s.stateLock.Unlock()
+		return nil
 	}
-	if client != nil {
-		if err := client.Close(); err != nil {
-			s.log.With("error", err).Error("Failed to close client")
+}
+
+func (s *sftpReader) newCodecAckFn(path string) service.AckFunc {
+	return func(ctx context.Context, aErr error) error {
+		s.stateLock.Lock()
+		defer s.stateLock.Unlock()
+
+		if err := s.pathProvider.Ack(ctx, path, aErr); err != nil {
+			s.log.With("error", err).Warnf("Failed to acknowledge path: %s", path)
+		}
+		if aErr != nil {
+			return nil
 		}
+
+		if s.deleteOnFinish {
+			if err := s.client.Remove(path); err != nil {
+				return fmt.Errorf("remove %v: %w", path, err)
+			}
+		}
+
+		time.Sleep(time.Millisecond * 100)
```

Comment:
Is this intentional? (If yes, a comment would be handy)

Reply:
Nope, removed